File: backend/app/routes/recurring_entries.py
"""
Recurring journal-entry engine — HTTP routes (Module A).

Prefix: /api/recurring-entries

Endpoints:
    GET    /                      list templates + last posting info
    POST   /                      create new template
    PUT    /{id}                  update template
    PATCH  /{id}/toggle           toggle is_active
    DELETE /{id}                  deactivate (soft delete)
    POST   /{id}/post             post a template for a period
    POST   /post-all              post all due templates for a period
    POST   /seed                  seed standard templates (admin)
    GET    /month-end-status      status for close-readiness panel
"""
from __future__ import annotations

from datetime import date as DateType
from decimal import Decimal
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from ..db import db_session
from ..services_auth import enforce_entity_code, require_role
from ..services_period_close import PeriodLockedError
from ..services_recurring_entries import (
    get_month_end_status,
    list_templates,
    post_due_templates,
    post_template,
    seed_standard_templates,
    set_template_active,
    upsert_template,
)
from ..services import get_entity_by_code

router = APIRouter(prefix="/api/recurring-entries", tags=["recurring-entries"])


def _parse_date(name: str, value: str) -> DateType:
    try:
        return DateType.fromisoformat(value)
    except ValueError as exc:
        raise HTTPException(
            status_code=400,
            detail=f"{name} must be YYYY-MM-DD, got {value!r}",
        ) from exc


# =========================================================================
# LIST
# =========================================================================

@router.get("")
def get_templates(
    entity_code: str = Query(...),
) -> dict[str, Any]:
    try:
        with db_session() as session:
            return list_templates(session, entity_code=entity_code)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# CREATE
# =========================================================================

class LineModel(BaseModel):
    line_number: int
    account_code: str
    direction: str  # 'debit' | 'credit'
    memo: str | None = None


class CreateTemplateRequest(BaseModel):
    entity_code: str
    name: str
    description: str | None = None
    calc_type: str  # 'fixed' | 'formula' | 'schedule'
    fixed_amount: float | None = None
    formula_expr: str | None = None
    schedule_source: str | None = None
    cadence: str = "monthly"
    posting_day: int = 1
    is_active: bool = False
    auto_post: bool | None = None
    notes: str | None = None
    lines: list[LineModel] = Field(default_factory=list)
    actor_email: str


@router.post("")
def post_create_template(
    body: CreateTemplateRequest,
    _user: dict = Depends(require_role("admin")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    try:
        with db_session() as session:
            return upsert_template(
                session,
                entity_code=body.entity_code,
                name=body.name,
                description=body.description,
                calc_type=body.calc_type,
                fixed_amount=Decimal(str(body.fixed_amount)) if body.fixed_amount is not None else None,
                formula_expr=body.formula_expr,
                schedule_source=body.schedule_source,
                cadence=body.cadence,
                posting_day=body.posting_day,
                is_active=body.is_active,
                auto_post=body.auto_post,
                notes=body.notes,
                lines=[l.model_dump() for l in body.lines],
                actor_email=body.actor_email,
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# UPDATE
# =========================================================================

class UpdateTemplateRequest(BaseModel):
    entity_code: str
    name: str
    description: str | None = None
    calc_type: str
    fixed_amount: float | None = None
    formula_expr: str | None = None
    schedule_source: str | None = None
    cadence: str = "monthly"
    posting_day: int = 1
    is_active: bool = False
    auto_post: bool | None = None
    notes: str | None = None
    lines: list[LineModel] | None = None
    actor_email: str


@router.put("/{template_id}")
def put_update_template(
    template_id: str,
    body: UpdateTemplateRequest,
    _user: dict = Depends(require_role("admin")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    try:
        with db_session() as session:
            return upsert_template(
                session,
                entity_code=body.entity_code,
                template_id=template_id,
                name=body.name,
                description=body.description,
                calc_type=body.calc_type,
                fixed_amount=Decimal(str(body.fixed_amount)) if body.fixed_amount is not None else None,
                formula_expr=body.formula_expr,
                schedule_source=body.schedule_source,
                cadence=body.cadence,
                posting_day=body.posting_day,
                is_active=body.is_active,
                auto_post=body.auto_post,
                notes=body.notes,
                lines=[l.model_dump() for l in body.lines] if body.lines is not None else None,
                actor_email=body.actor_email,
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# TOGGLE active
# =========================================================================

class ToggleRequest(BaseModel):
    entity_code: str
    is_active: bool


@router.patch("/{template_id}/toggle")
def patch_toggle(
    template_id: str,
    body: ToggleRequest,
    _user: dict = Depends(require_role("admin")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    try:
        with db_session() as session:
            return set_template_active(
                session,
                entity_code=body.entity_code,
                template_id=template_id,
                is_active=body.is_active,
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# DELETE (soft deactivate)
# =========================================================================

class DeactivateRequest(BaseModel):
    entity_code: str


@router.delete("/{template_id}")
def delete_template(
    template_id: str,
    body: DeactivateRequest,
    _user: dict = Depends(require_role("admin")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    try:
        with db_session() as session:
            return set_template_active(
                session,
                entity_code=body.entity_code,
                template_id=template_id,
                is_active=False,
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# POST (trigger posting for a period)
# =========================================================================

class PostRequest(BaseModel):
    entity_code: str
    period_end: str = Field(..., examples=["2026-03-31"])
    actor_email: str
    dry_run: bool = False


@router.post("/{template_id}/post")
def post_trigger(
    template_id: str,
    body: PostRequest,
    _user: dict = Depends(require_role("bookkeeper")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    period_end = _parse_date("period_end", body.period_end)
    try:
        with db_session() as session:
            return post_template(
                session,
                entity_code=body.entity_code,
                template_id=template_id,
                period_end=period_end,
                actor_email=body.actor_email,
                dry_run=body.dry_run,
            )
    except PeriodLockedError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        # Already-posted or balance-guard errors
        status = 409 if "already posted" in str(exc) or "unapproved draft" in str(exc) else 400
        raise HTTPException(status_code=status, detail=str(exc)) from exc


# =========================================================================
# POST-ALL (bulk trigger all due)
# =========================================================================

class PostAllRequest(BaseModel):
    entity_code: str
    period_end: str
    actor_email: str


@router.post("/post-all")
def post_all_due(
    body: PostAllRequest,
    _user: dict = Depends(require_role("bookkeeper")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    period_end = _parse_date("period_end", body.period_end)
    try:
        with db_session() as session:
            return post_due_templates(
                session,
                entity_code=body.entity_code,
                period_end=period_end,
                actor_email=body.actor_email,
            )
    except PeriodLockedError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# SEED standard templates
# =========================================================================

class SeedRequest(BaseModel):
    entity_code: str


@router.post("/seed")
def post_seed(
    body: SeedRequest,
    _user: dict = Depends(require_role("admin")),
) -> dict[str, Any]:
    enforce_entity_code(_user, body.entity_code)
    try:
        with db_session() as session:
            return seed_standard_templates(session, entity_code=body.entity_code)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# =========================================================================
# MONTH-END STATUS
# =========================================================================

@router.get("/month-end-status")
def get_me_status(
    entity_code: str = Query(...),
    period_end: str = Query(...),
) -> dict[str, Any]:
    period_end_d = _parse_date("period_end", period_end)
    try:
        with db_session() as session:
            entity = get_entity_by_code(session, entity_code)
            if not entity:
                raise HTTPException(status_code=404, detail=f"Entity not found: {entity_code}")
            return get_month_end_status(
                session, entity_id=entity["id"], period_end=period_end_d
            )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc

File: frontend/lib/api/recurring_entries.ts
import { api } from './client';

export interface RecurringLine {
  line_number: number;
  account_code: string;
  direction: 'debit' | 'credit';
  memo?: string | null;
}

export interface RecurringTemplate {
  id: string;
  name: string;
  description: string | null;
  standard_key: string | null;
  calc_type: 'fixed' | 'formula' | 'schedule';
  fixed_amount: string | null;
  formula_expr: string | null;
  schedule_source: string | null;
  cadence: 'monthly' | 'on_close' | 'annual';
  posting_day: number;
  is_active: boolean;
  auto_post: boolean;
  last_posted_at: string | null;
  last_posted_period_end: string | null;
  total_postings: number;
  notes: string | null;
  lines: RecurringLine[];
}

export interface MonthEndStatusItem {
  template_id: string;
  name: string;
  status: 'posted' | 'pending';
  amount: string | null;
  journal_batch_id: string | null;
  auto_posted?: boolean;
}

export async function listRecurringTemplates(
  entityCode: string,
): Promise<{ templates: RecurringTemplate[]; count: number }> {
  const res = await api.get('/api/recurring-entries', {
    params: { entity_code: entityCode },
  });
  return res.data;
}

export async function createRecurringTemplate(input: {
  entity_code: string;
  name: string;
  description?: string | null;
  calc_type: string;
  fixed_amount?: number | null;
  formula_expr?: string | null;
  schedule_source?: string | null;
  cadence?: string;
  posting_day?: number;
  is_active?: boolean;
  auto_post?: boolean | null;
  notes?: string | null;
  lines?: RecurringLine[];
  actor_email: string;
}): Promise<{ id: string }> {
  const res = await api.post('/api/recurring-entries', input);
  return res.data;
}

export async function updateRecurringTemplate(
  templateId: string,
  input: {
    entity_code: string;
    name: string;
    description?: string | null;
    calc_type: string;
    fixed_amount?: number | null;
    formula_expr?: string | null;
    schedule_source?: string | null;
    cadence?: string;
    posting_day?: number;
    is_active?: boolean;
    auto_post?: boolean | null;
    notes?: string | null;
    lines?: RecurringLine[] | null;
    actor_email: string;
  },
): Promise<{ id: string }> {
  const res = await api.put(`/api/recurring-entries/${templateId}`, input);
  return res.data;
}

export async function toggleRecurringTemplate(
  templateId: string,
  entityCode: string,
  isActive: boolean,
): Promise<{ id: string; is_active: boolean }> {
  const res = await api.patch(`/api/recurring-entries/${templateId}/toggle`, {
    entity_code: entityCode,
    is_active: isActive,
  });
  return res.data;
}

export async function deleteRecurringTemplate(
  templateId: string,
  entityCode: string,
): Promise<{ id: string; is_active: boolean }> {
  const res = await api.delete(`/api/recurring-entries/${templateId}`, {
    data: { entity_code: entityCode },
  });
  return res.data;
}

export async function postRecurringTemplate(input: {
  entity_code: string;
  template_id: string;
  period_end: string;
  actor_email: string;
  dry_run?: boolean;
}): Promise<{
  dry_run: boolean;
  journal_batch_id?: string;
  batch_status?: string;
  grand_total: string;
  journal_lines: Array<{
    line_number: number;
    account_code: string;
    debit_amount: string;
    credit_amount: string;
    memo: string;
  }>;
}> {
  const { template_id, ...body } = input;
  const res = await api.post(`/api/recurring-entries/${template_id}/post`, body);
  return res.data;
}

export async function postAllDueTemplates(input: {
  entity_code: string;
  period_end: string;
  actor_email: string;
}): Promise<{
  posted_count: number;
  skipped_count: number;
  failed_count: number;
  posted: Array<{ template_id: string; journal_batch_id: string; grand_total: string }>;
  skipped: Array<{ template_id: string; reason: string }>;
  failed: Array<{ template_id: string; error: string }>;
}> {
  const res = await api.post('/api/recurring-entries/post-all', input);
  return res.data;
}

export async function seedRecurringTemplates(entityCode: string): Promise<{
  inserted: number;
  skipped: number;
}> {
  const res = await api.post('/api/recurring-entries/seed', { entity_code: entityCode });
  return res.data;
}

export async function getRecurringMonthEndStatus(
  entityCode: string,
  periodEnd: string,
): Promise<{
  status: string;
  summary: string;
  items: MonthEndStatusItem[];
  all_posted: boolean;
}> {
  const res = await api.get('/api/recurring-entries/month-end-status', {
    params: { entity_code: entityCode, period_end: periodEnd },
  });
  return res.data;
}
